File: smash-ranks-android/app/src/test/java/com/garpr/android/preferences/persistent/PersistentIntegerPreferenceTest.java
package com.garpr.android.preferences.persistent;

import com.garpr.android.BaseTest;
import com.garpr.android.BuildConfig;
import com.garpr.android.preferences.KeyValueStore;
import com.garpr.android.preferences.Preference;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import javax.inject.Inject;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertNull;

@RunWith(RobolectricTestRunner.class)
@Config(constants = BuildConfig.class)
public class PersistentIntegerPreferenceTest extends BaseTest {

    @Inject
    KeyValueStore mKeyValueStore;


    @Before
    @Override
    public void setUp() throws Exception {
        super.setUp();
        getTestAppComponent().inject(this);
    }

    @Test
    public void testAddListener() throws Exception {
        final Preference<Integer> preference = new PersistentIntegerPreference("integer", null,
                mKeyValueStore);

        final int[] array = { 0 };

        final Preference.OnPreferenceChangeListener<Integer> listener =
                new Preference.OnPreferenceChangeListener<Integer>() {
            @Override
            public void onPreferenceChange(final Preference<Integer> preference) {
                // noinspection ConstantConditions
                array[0] = preference.get();
            }
        };

        preference.addListener(listener);

        preference.set(10);
        assertEquals(10, array[0]);

        preference.set(20);
        assertEquals(20, array[0]);
    }

    @Test
    public void testDelete() throws Exception {
        final Preference<Integer> preference = new PersistentIntegerPreference("integer", -47,
                mKeyValueStore);

        assertEquals(preference.get(), Integer.valueOf(-47));

        preference.delete();
        assertEquals(preference.get(), Integer.valueOf(-47));

        preference.set(100);
        assertEquals(preference.get(), Integer.valueOf(100));

        preference.delete();
        assertEquals(preference.get(), Integer.valueOf(-47));
    }

    @Test
    public void testGetAndSet() throws Exception {
        final Preference<Integer> preference = new PersistentIntegerPreference("integer", 900,
                mKeyValueStore);

        assertEquals(preference.get(), Integer.valueOf(900));

        preference.set(0);
        assertEquals(preference.get(), Integer.valueOf(0));

        preference.set((Integer) null);
        assertEquals(preference.get(), Integer.valueOf(900));
        assertEquals(preference.get(), preference.getDefaultValue());

        preference.set(-20);
        assertEquals(preference.get(), Integer.valueOf(-20));
        assertNotEquals(preference.get(), preference.getDefaultValue());
    }

    @Test
    public void testNonNullGetDefaultValue() throws Exception {
        final Preference<Integer> preference = new PersistentIntegerPreference("integer", 128,
                mKeyValueStore);

        assertEquals(preference.getDefaultValue(), Integer.valueOf(128));

        preference.set(Integer.MAX_VALUE);
        assertEquals(preference.getDefaultValue(), Integer.valueOf(128));
    }

    @Test
    public void testNullGetDefaultValue() throws Exception {
        final Preference<Integer> preference = new PersistentIntegerPreference("integer", null,
                mKeyValueStore);

        assertNull(preference.getDefaultValue());

        preference.set(20);
        assertNull(preference.getDefaultValue());
    }

    @Test
    public void testRemoveListener() throws Exception {
        final Preference<Integer> preference = new PersistentIntegerPreference("integer", null,
                mKeyValueStore);

        final int[] array = { 0 };

        final Preference.OnPreferenceChangeListener<Integer> listener =
                new Preference.OnPreferenceChangeListener<Integer>() {
            @Override
            public void onPreferenceChange(final Preference<Integer> preference) {
                // noinspection ConstantConditions
                array[0] = preference.get();
            }
        };

        preference.addListener(listener);

        preference.set(10);
        assertEquals(10, array[0]);

        preference.set(20);
        assertEquals(20, array[0]);

        preference.removeListener(listener);

        preference.set(100);
        assertNotEquals(100, array[0]);
    }

    @Test
    public void testSetAndGet() throws Exception {
        final Preference<Integer> preference = new PersistentIntegerPreference("integer", null,
                mKeyValueStore);

        preference.set(5);
        assertEquals(preference.get(), Integer.valueOf(5));

        preference.set(10);
        assertEquals(preference.get(), Integer.valueOf(10));

        preference.set((Integer) null);
        assertEquals(preference.get(), null);

        preference.set(Integer.MIN_VALUE);
        assertEquals(preference.get(), Integer.valueOf(Integer.MIN_VALUE));
    }

}

File: smash-ranks-android/app/src/test/java/com/garpr/android/preferences/persistent/PersistentIntegerPreferenceTest.kt
package com.garpr.android.preferences.persistent

import com.garpr.android.BaseTest
import com.garpr.android.data.models.Optional
import com.garpr.android.preferences.KeyValueStore
import com.garpr.android.preferences.KeyValueStoreProvider
import com.garpr.android.preferences.Preference
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.koin.test.inject
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class PersistentIntegerPreferenceTest : BaseTest() {

    protected val keyValueStoreProvider: KeyValueStoreProvider by inject()

    private lateinit var keyValueStore: KeyValueStore

    companion object {
        private const val TAG = "PersistentIntegerPreferenceTest"
    }

    @Before
    override fun setUp() {
        super.setUp()

        keyValueStore = keyValueStoreProvider.getKeyValueStore(TAG)
    }

    @Test
    fun testExistsWithDefaultValue() {
        val preference: Preference<Int> = PersistentIntegerPreference(
                key = "integer",
                defaultValue = 1989,
                keyValueStore = keyValueStore
        )
        assertTrue(preference.exists)
    }

    @Test
    fun testExistsWithNullDefaultValue() {
        val preference: Preference<Int> = PersistentIntegerPreference(
                key = "integer",
                defaultValue = null,
                keyValueStore = keyValueStore
        )
        assertFalse(preference.exists)
    }

    @Test
    fun testGetAndSetAndDeleteAndExistsWithDefaultValue() {
        val preference: Preference<Int> = PersistentIntegerPreference(
                key = "integer",
                defaultValue = 900,
                keyValueStore = keyValueStore
        )
        assertEquals(900, preference.get())
        assertTrue(preference.exists)

        preference.set(0)
        assertEquals(0, preference.get())
        assertTrue(preference.exists)

        preference.delete()
        assertEquals(900, preference.get())
        assertTrue(preference.exists)

        preference.set(-20)
        assertEquals(-20, preference.get())
        assertTrue(preference.exists)
    }

    @Test
    fun testGetAndSetAndDeleteAndExistsWithNullDefaultValue() {
        val preference: Preference<Int> = PersistentIntegerPreference(
                key = "integer",
                defaultValue = null,
                keyValueStore = keyValueStore
        )
        assertNull(preference.get())
        assertFalse(preference.exists)

        preference.set(88)
        assertEquals(88, preference.get())
        assertTrue(preference.exists)

        preference.delete()
        assertNull(preference.get())
        assertFalse(preference.exists)

        preference.set(-20)
        assertEquals(-20, preference.get())
        assertTrue(preference.exists)
    }

    @Test
    fun testKey() {
        val preference: Preference<Int> = PersistentIntegerPreference(
                key = "integer",
                defaultValue = null,
                keyValueStore = keyValueStore
        )
        assertEquals("integer", preference.key)
    }

    @Test
    fun testObserveWithDefaultValue() {
        val preference: Preference<Int> = PersistentIntegerPreference(
                key = "integer",
                defaultValue = -31,
                keyValueStore = keyValueStore
        )

        var value: Optional<Int>? = null

        preference.observable.subscribe {
            value = it
        }

        assertNull(value)

        preference.set(25)
        assertNotNull(value)
        assertEquals(25, value?.item)

        preference.set(10)
        assertNotNull(value)
        assertEquals(10, value?.item)

        preference.delete()
        assertNotNull(value)
        assertEquals(-31, value?.item)
    }

    @Test
    fun testObserveWithNullDefaultValue() {
        val preference: Preference<Int> = PersistentIntegerPreference(
                key = "integer",
                defaultValue = null,
                keyValueStore = keyValueStore
        )

        var value: Optional<Int>? = null

        preference.observable.subscribe {
            value = it
        }

        assertNull(value)

        preference.set(11)
        assertNotNull(value)
        assertEquals(11, value?.item)

        preference.set(168)
        assertNotNull(value)
        assertEquals(168, value?.item)

        preference.delete()
        assertNotNull(value)
        assertEquals(false, value?.isPresent)
    }

    @Test
    fun testSetWithNullCausesDelete() {
        val preference: Preference<Int> = PersistentIntegerPreference(
                key = "integer",
                defaultValue = null,
                keyValueStore = keyValueStore
        )

        preference.set(-73)
        assertTrue(preference.exists)

        preference.set(null)
        assertFalse(preference.exists)
    }

}
